src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,518 @@ extern "C" {
+	}
+
+	if (IsA(node, SubLink))
+	{
+		SubLink *sublink = (SubLink *) node;
+
+		context->m_current_query_level++;
+		GetVarsWithoutTleWalker((Node *) sublink->subselect, context);
+		context->m_current_query_level--;
+
+		return false;
+	}
+
+	// recurse into query structure
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
+		GetVarsWithoutTleWalker((Node *) query->targetList, context);
+
+		List *rtable = query->rtable;
+		ListCell *lc = NULL;
+		ForEach(lc, rtable)
+		{
+			RangeTblEntry *rte = (RangeTblEntry *) lfirst(lc);
+
+			if (rte != NULL && RTE_SUBQUERY == rte->rtekind)
+			{
+				Query *subquery = rte->subquery;
+				context->m_current_query_level++;
+				GetVarsWithoutTleWalker((Node *) subquery, context);
+				context->m_current_query_level--;
+			}
+		}
+	}
+
+	return gpdb::WalkExpressionTree(
+		node, (ExprWalkerFn) GetVarsWithoutTleWalker, (void *) context);
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::FixGroupDependentTargets
+//
+//	@doc:
+//		Updates the query structure in case there are target list entries
+//		with functional dependency on columns in groupClause.
+//---------------------------------------------------------------------------
+void
+CQueryMutators::FixGroupDependentTargets(Query *query)
+{
+	// If there is a functional dependency proved for a relation on a set of
+	// grouping columns, then it is valid to have a column in the target list
+	// even if it is not listed explicitly in the groupClause. To make ORCA
+	// correctly process such cases, we add all such target list entries into
+	// groupClause's grouping sets, where the functional dependency exists.
+	// If there is such functional dependency existing, the OID of
+	// Primary Key constraint resides in the constraintDeps list. As the
+	// constraintDeps list contains only Primary Key OIDs, we skip checks of
+	// constraint type in all code below.
+	// Refer to function check_functional_grouping for more details.
+	if (0 != gpdb::ListLength(query->constraintDeps))
+	{
+		// 1. Extract such columns from targetlist expressions, and if they do
+		// not have a relevant target entry, add resjunk target list entries for
+		// them.
+		// 2. Store current unique TargetEntry references in groupClause -
+		// they will be needed at step 4.
+		// 3. Update all grouping sets that contain Primary Key
+		// 4. Update arguments of GROUPING functions, because they could be
+		// shifted after step 3.
+
+		// Step 1.
+		// If there is an expression with a functionally dependent var, at this
+		// point it may not have a relevant target list entry (as it was not
+		// explicitly listed in groupClause).
+		// For all such vars we add resjunc target list entries into targetList.
+		SContextFixGroupDependentTargets ctx_fix_dependent_targets(query);
+		GetVarsWithoutTleWalker((Node *) query->targetList,
+								&ctx_fix_dependent_targets);
+		query->targetList = gpdb::ListConcat(
```

Comment:
>3. Update all grouping sets that contain the primary key - add the functionally
dependent columns from the target list. In case column from target list
expression do not have a relevant target list entry, add a resjunk target list
entry for it.

remove second sentence as it repeats the sentence from paragraph 1

Reply:
done